packages/flutter/test/material/tabs_utils.dart

```diff
@@ -0,0 +1,239 @@
+}
+
+TabController getTabController({required int length, required TickerProvider vsync, int initialIndex = 0, Duration? animationDuration}) {
+  final TabController result = TabController(length: length, vsync: vsync, initialIndex: initialIndex, animationDuration: animationDuration);
+  addTearDown(result.dispose);
+  return result;
+}
+
+typedef TabControllerFrameBuilder = Widget Function(BuildContext context, TabController controller);
+
+class TabStateMarker extends StatefulWidget {
+  const TabStateMarker({ super.key, this.child });
+
+  final Widget? child;
+
+  @override
+  State<TabStateMarker> createState() => TabStateMarkerState();
+}
+
+class TabStateMarkerState extends State<TabStateMarker> {
+  String? marker;
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.child != null) {
+      return widget.child!;
+    }
+    return Container();
+  }
+}
+
+class NestedTabBarContainer extends StatelessWidget {
+  const NestedTabBarContainer({
+    super.key,
+    this.tabController,
+  });
+
+  final TabController? tabController;
+
+  @override
+  Widget build(BuildContext context) {
+    return ColoredBox(
+      color: Colors.blue,
+      child: Column(
+        children: <Widget>[
+          TabBar(
+            controller: tabController,
+            tabs: const <Tab>[
+              Tab(text: 'Yellow'),
+              Tab(text: 'Grey'),
+            ],
+          ),
+          Expanded(
+            child: TabBarView(
+              controller: tabController,
+              children: <Widget>[
+                Container(color: Colors.yellow),
+                Container(color: Colors.grey),
+              ],
+            ),
+          ),
+        ],
+      ),
+    );
+  }
+}
+
+class TabControllerFrame extends StatefulWidget {
+  const TabControllerFrame({
+    super.key,
+    required this.length,
+    this.initialIndex = 0,
+    required this.builder,
+  });
+
+  final int length;
+  final int initialIndex;
+  final TabControllerFrameBuilder builder;
+
+  @override
+  TabControllerFrameState createState() => TabControllerFrameState();
+}
+
+class TabControllerFrameState extends State<TabControllerFrame> with SingleTickerProviderStateMixin {
+  late TabController _controller;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = TabController(
+      vsync: this,
+      length: widget.length,
+      initialIndex: widget.initialIndex,
+    );
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return widget.builder(context, _controller);
+  }
+}
+
+class TabIndicatorRecordingCanvas extends TestRecordingCanvas {
+  TabIndicatorRecordingCanvas(this.indicatorColor);
+
+  final Color indicatorColor;
+  late Rect indicatorRect;
+
+  @override
+  void drawLine(Offset p1, Offset p2, Paint paint) {
+    // Assuming that the indicatorWeight is 2.0, the default.
+    const double indicatorWeight = 2.0;
```

Comment:
Is this different in M2 versus M3?

Reply:
i'll update the tests in the follow up PRs and verify this.

Reply:
Technically, indicatorWeight should `2` for tab indicator and `3` in label indicator for Material  3.